lib/prototypes/abstract.js

```diff
@@ -14,16 +14,22 @@ class AbstractType {
     }
   }
 
+  #root;
+
   constructor(def, preprocessor) {
-    this.root = preprocessor.root;
+    this.#root = preprocessor.root;
     const { formatters } = AbstractType;
     for (const [key, value] of Object.entries(def)) {
       if (key === 'type' || key === this.type) continue;
       if (formatters[key]) this[key] = formatters[key](value);
       else this[key] = value;
     }
     this.construct(def, preprocessor);
-    if (this.type) this.root.references.add(this.type);
+    if (this.type) this.#root.references.add(this.type);
+  }
+
+  get root() {
+    return this.#root;
```

Comment:
Why do we need private `#root` if it will be accessible as public `root` with getter?